freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaCreationService.java

```diff
@@ -116,8 +128,15 @@ public DescribeFreeIpaResponse launchFreeIpa(CreateFreeIpaRequest request, Strin
                 Stack savedStack = stackService.save(stack);
                 ImageSettingsRequest imageSettingsRequest = request.getImage();
 
-                // JSA todo populate CCM parameters
                 CcmParameters ccmParameters = null;
+                if ((ccmParameterSupplier != null) && Boolean.TRUE.equals(stack.getUseCcm())) {
+                    int gatewayPort = Optional.ofNullable(stack.getGatewayport()).orElse(ServiceFamilies.GATEWAY.getDefaultPort());
+                    Map<KnownServiceIdentifier, Integer> tunneledServicePorts = Collections.singletonMap(KnownServiceIdentifier.GATEWAY, gatewayPort);
+                    // JSA TODO Use stack ID or something else instead?
+                    String keyId = StringUtils.right(stack.getResourceCrn(), CCM_KEY_ID_LENGTH);
```

Comment:
The KeyId is what we use to look up the tunnel when we want to use it, right? shouldn't we save it as part of the stack model so we can retrieve it later?

Reply:
We should be looking the tunnels up using the provider-specific instance ID. As far as I know, once we pass the key ID to the instance for key decipherment, we don't need it again on the CP side.